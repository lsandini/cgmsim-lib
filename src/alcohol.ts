import { TreatmentDelta, Treatment, TreatmentDrug, GenderType } from './Types';
import { getDrugActivity } from './drug';
import logger, { getDeltaMinutes, getTreatmentActivity } from './utils';

function getAlcoholActivity(
	gender: 'Male' | 'Female',
	weightKg: number,
	peakMin: number,
	durationMin: number,
	timeMin: number, //minutesAgo
	units: number,
): number {
	const peakAlcoholAmountMin = 60; //min
	const r = gender === 'Male' ? 0.68 : 0.55;
	const eliminationRate = gender === 'Male' ? 0.016 / 60 : 0.018 / 60; //g/100ml/min
	const peakAmount = ((units * 12) / (weightKg * 1000 * r)) * 100; //g/100ml
	const washoutDuration = peakAlcoholAmountMin + peakAmount / eliminationRate; //min
	const unitsWeighted = units * 12 * (80 / weightKg);
	if (washoutDuration < timeMin) {
		return (
			getTreatmentActivity(
				peakMin,
				durationMin,
				timeMin - washoutDuration,
				unitsWeighted,
			) / 35
		);
	}
	return 0;
}

export const computeAlcoholActivity = (
	treatments: TreatmentDelta[],
	weight: number,
	gender: GenderType,
) => {
	const treatmentsActivity = treatments.map((e) => {
		const minutesAgo = e.minutesAgo;
		const units = e.units;
		const activity = getAlcoholActivity(
			gender,
			weight,
			e.peak,
			e.duration,
			minutesAgo,
			units,
		);
		return activity;
	});
	logger.debug('these are the last Alcohol: %o', treatmentsActivity);
	const resultAct = treatmentsActivity.reduce((tot, activity) => {
		return tot + activity;
	}, 0);
	return resultAct;
};

export default function (
	treatments: TreatmentDrug[],
	weight: number,
	gender: GenderType,
): number {
	//Find Alcohol boluses

	const lastALC = getDrugActivity(treatments, weight, 'ALC');
	const activityALC =
		lastALC.length > 0 ? computeAlcoholActivity(lastALC, weight, gender) : 0;
	logger.debug('these are the last ALC: %o', { lastALC, activityALC });

	return activityALC / 2;
}
