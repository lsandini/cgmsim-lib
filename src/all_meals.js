const moment = require('moment');
module.exports = function(treatments = []) {
  const now = moment();
  const meals = treatments
    .filter(e => e.carbs && moment(e.mills).diff(now, 'minutes') > -360)
    .map(e => ({
      ...e,
      time:  (Date.now() - moment(e.mills).valueOf()) / (1000 * 60)
    }));

  console.log('Last 6 hours meal:', meals);
  return meals;
};
