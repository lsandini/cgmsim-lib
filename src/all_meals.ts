import logger from './utils';
import { Treatment, TreatmentDelta } from './Types';
import { getDeltaMinutes } from './utils';

//const logger = pino();

export default function (treatments: Treatment[]): TreatmentDelta[] {

	const meals = treatments
		.filter(e => e.carbs && getDeltaMinutes(e.mills) > 360)
		.map(e => ({
			...e,
			minutesAgo: getDeltaMinutes(e.mills),
		}));

	logger.info('Last 6 hours meal:', meals);
	return meals;
}

