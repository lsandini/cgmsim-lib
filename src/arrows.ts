import logger from './utils';
import { Sgv, Direction } from './Types';

//const logger = pino();


export default function (sgvs: Pick<Sgv, 'sgv'>[]): { sgvdir: number; direction: Direction }[] {
	const arrows = [];
	if (sgvs && sgvs.length >= 4) {
		const sgvdir1 = sgvs[0].sgv - sgvs[1].sgv;
		const sgvdir2 = sgvs[1].sgv - sgvs[2].sgv;
		const sgvdir3 = sgvs[2].sgv - sgvs[3].sgv;
		const sgvdir15min = (sgvdir1 + sgvdir2 + sgvdir3) / 3;
		logger.info('this is the mean SGV 5 min variation in the last 15 minutes: %o', sgvdir15min, 'mg/dl');

		if (sgvdir15min < -10) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'DoubleDown'
			});

		} else if (sgvdir15min < -6) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'SingleDown'
			});

		} else if (sgvdir15min < -2) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'FortyFiveDown'
			});

		} else if (sgvdir15min < 2) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'Flat'
			});

		} else if (sgvdir15min < 6) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'FortyFiveUp'
			});

		} else if (sgvdir15min < 10) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'SingleUp'
			});

		} else if (sgvdir15min >= 10) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'DoubleUp'
			});
		}

	} else {
		arrows.push({
			sgvdir: 0,
			direction: 'Flat'
		});
	}
	logger.info(arrows);

	return arrows;
}