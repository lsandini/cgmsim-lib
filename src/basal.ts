import { TreatmentDelta, Treatment } from './Types';
import logger, { getDeltaMinutes, getInsulinActivity } from './utils';

export const peakBasal = {
	GLA: (duration: number) => duration / 2.5,
	DET: (duration: number) => duration / 3,
	TOU: (duration: number) => duration / 2.5,
	DEG: (duration: number) => duration / 3,
};
export const durationBasal = {
	GLA: (insulin: number, weight: number) => (22 + (12 * insulin) / weight) * 60,
	DET: (insulin: number, weight: number) => (14 + (24 * insulin) / weight) * 60,
	TOU: (insulin: number, weight: number) => (24 + (14 * insulin) / weight) * 60,
	DEG: () => 42 * 60,
};

export const computeBasalActivity = (
	treatments: (TreatmentDelta & { duration: number; peak: number })[]
) => {
	// activities be expressed as U/min !!!
	const treatmentsActivity = treatments.map((e) => {
		const minutesAgo = e.minutesAgo;
		const insulin = e.insulin;
		const activity = getInsulinActivity(
			e.peak,
			e.duration,
			minutesAgo,
			insulin
		);
		return activity;
	});
	logger.debug('these are the last Slow INSULINS: %o', treatmentsActivity);
	const resultAct = treatmentsActivity.reduce((tot, activity) => {
		return tot + activity;
	}, 0);
	return resultAct;
};

export default function (treatments: Treatment[], weight: number): number {
	//Find basal boluses
	const basals =
		treatments && treatments.length
			? treatments
					.filter((e) => e.notes)
					.map((e) => {
						const lastIndexEmptySpace = e.notes.lastIndexOf(' ');
						logger.debug(
							'tou %o',
							parseInt(e.notes.slice(lastIndexEmptySpace), 10)
						);
						return {
							...e,
							minutesAgo: getDeltaMinutes(e.created_at),
							drug: e.notes.slice(0, 3),
							// insulin: parseInt(e.notes.slice(-2))
							insulin: parseInt(e.notes.slice(lastIndexEmptySpace), 10) || 0,
						};
					})
					.filter((e) => e.minutesAgo >= 0)
			: [];

	const lastBasals = basals.filter(function (e) {
		return e.minutesAgo <= 45 * 60; // keep only the basals from the last 45 hours
	});

	const lastGLA = lastBasals
		.filter((e) => {
			return (
				e.drug === 'gla' ||
				e.drug === 'Gla' ||
				e.drug === 'lan' ||
				e.drug === 'Lan'
			); // keep only the glas from the last 45 hours
		})
		.map((e) => {
			const duration = durationBasal.GLA(e.insulin, weight);
			const peak = peakBasal.GLA(duration);
			return {
				...e,
				duration,
				peak,
			};
		});
	const activityGLA = lastGLA.length > 0 ? computeBasalActivity(lastGLA) : 0;
	logger.debug('these are the last GLA: %o', { lastGLA, activityGLA });

	const lastDET = lastBasals
		.filter(function (e) {
			return (
				e.drug === 'det' ||
				e.drug === 'Det' ||
				e.drug === 'lev' ||
				e.drug === 'Lev'
			); // keep only the dets from the last 45 hours
		})
		.map((e) => {
			const duration = durationBasal.DET(e.insulin, weight);
			const peak = peakBasal.DET(duration);
			return {
				...e,
				duration,
				peak,
			};
		});

	const activityDET = lastDET.length ? computeBasalActivity(lastDET) : 0;
	logger.debug('these are the last DET: %o', { lastDET, activityDET });

	const lastTOU = lastBasals
		.filter(function (e) {
			return e.drug === 'tou' || e.drug === 'Tou'; // keep only the tous from the last 45 hours
		})
		.map((e) => {
			const duration = durationBasal.TOU(e.insulin, weight);
			const peak = peakBasal.TOU(duration);
			return {
				...e,
				duration,
				peak,
			};
		});
	const activityTOU = lastTOU.length ? computeBasalActivity(lastTOU) : 0;
	logger.debug('these are the last TOU: %o', { lastTOU, activityTOU });

	const lastDEG = lastBasals
		.filter(function (e) {
			return (
				e.drug === 'deg' ||
				e.drug === 'Deg' ||
				e.drug === 'tre' ||
				e.drug === 'Tre'
			); // keep only the degs from the last 45 hours
		})
		.map((e) => {
			const duration = durationBasal.DEG();
			const peak = peakBasal.DEG(duration);
			return {
				...e,
				duration,
				peak,
			};
		});
	const activityDEG = lastDEG.length ? computeBasalActivity(lastDEG) : 0;
	logger.debug('these are the last deg: %o', { lastDEG, activityDEG });


	return activityDEG + activityDET + activityGLA + activityTOU ;
}
