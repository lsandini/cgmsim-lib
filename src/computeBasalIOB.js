const moment = require('moment');
module.exports = function(treatments) {

	//Find basal boluses
    const basal = treatments && treatments.length ? treatments.filter(e => e.notes && e.insulin && e.drug !== 'hum')
        .map(e => ({
            time: e.created_at,
            notes: e.notes,
            insulin: e.insulin,
            empty_space: e.notes.indexOf(' ')
        })) : [];

    const basals = basal.map(entry => ({
        ...entry,
        time: moment(entry.time).valueOf()
    }));
    const timeSinceBasalMin = basals.map(entry => ({
        ...entry,
        time: (Date.now() - moment(entry.time).valueOf()) / (1000 * 60 * 60),
        drug: entry.notes.slice(0, 3),
        //		insulin: parseInt(entry.notes.slice(entry.empty_space), 10)
    }));

    const lastBasals = timeSinceBasalMin.filter(function(e) {
        return e.time <= 45; // keep only the basals from the last 45 hours
    });

    const lastGLA = lastBasals.filter(function(e) {
        return e.drug === 'gla' || e.drug === 'Gla' || e.drug === 'lan' || e.drug === 'Lan'; // keep only the glas from the last 45 hours
    });

    const lastDET = lastBasals.filter(function(e) {
        return e.drug === 'det' || e.drug === 'Det' || e.drug === 'lev' || e.drug === 'Lev'; // keep only the dets from the last 45 hours
    });

    const lastTOU = lastBasals.filter(function(e) {
        return e.drug === 'tou' || e.drug === 'Tou'; // keep only the tous from the last 45 hours
    });

    const lastDEG = lastBasals.filter(function(e) {
        return e.drug === 'deg' || e.drug === 'Deg' || e.drug === 'tre' || e.drug === 'Tre'; // keep only the degs from the last 45 hours
    });

    return {
        lastDET,
        lastGLA,
        lastTOU,
        lastDEG,
    };
};