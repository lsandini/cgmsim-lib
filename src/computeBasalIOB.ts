import { TreatmentDelta, Treatment } from './Types'
import logger, { getDeltaMinutes, getInsulinActivity } from './utils';


export const computeBasalActivity = (treatments: (TreatmentDelta & { duration: number, peak: number })[]) => {
	// activities be expressed as U/min !!!
	const treatmentsActivity = treatments.map(e => {
		const minutesAgo = e.minutesAgo;
		const insulin = e.insulin;
		return getInsulinActivity(e.peak, e.duration, minutesAgo, insulin);		
	});
	logger.info('these are the last Slow INSULINS: %o', treatmentsActivity);
	const resultAct = treatmentsActivity.reduce((tot, activity) => {
		return tot + activity;
	}, 0);
	return resultAct
}

export default function (treatments: Treatment[], weight: number): number {
	//Find basal boluses
	const basals = treatments && treatments.length ?
		treatments.filter(e => e.notes)
			.map(e => {
				const lastIndexEmptySpace = e.notes.lastIndexOf(' ');
				return {
					...e,
					minutesAgo: getDeltaMinutes(e.created_at),
					drug: e.notes.slice(0, 3),
					// insulin: parseInt(e.notes.slice(-2)) 
					insulin: parseInt(e.notes.slice(lastIndexEmptySpace), 10)
				}
			}) : [];

	const lastBasals = basals.filter(function (e) {
		return e.minutesAgo <= 45 * 60; // keep only the basals from the last 45 hours
	});

	const lastGLA = lastBasals
		.filter((e) => {
			return e.drug === 'gla' || e.drug === 'Gla' || e.drug === 'lan' || e.drug === 'Lan'; // keep only the glas from the last 45 hours
		}).map(e => {
			const duration = (22 + (12 * e.insulin / weight)) * 60;
			const peak = (duration / 2.5);
			return {
				...e,
				duration,
				peak,
			}
		});
	logger.debug('these are the last GLA: %o', lastGLA);
	const activityGLA = computeBasalActivity(lastGLA)
	logger.info('these is the total GLA activity: %o', activityGLA);

	const lastDET = lastBasals.filter(function (e) {
		return e.drug === 'det' || e.drug === 'Det' || e.drug === 'lev' || e.drug === 'Lev'; // keep only the dets from the last 45 hours
	}).map(e => {
		const duration = (14 + (24 * e.insulin / weight)) * 60;
		const peak = (duration / 3);
		return {
			...e,
			duration,
			peak,
		}
	});
	logger.debug('these are the last DET: %o', lastDET);
	const activityDET = computeBasalActivity(lastDET)
	logger.info('these is the total DET activity: %o', activityDET);

	const lastTOU = lastBasals.filter(function (e) {
		return e.drug === 'tou' || e.drug === 'Tou'; // keep only the tous from the last 45 hours
	})
	.map(e => {
		const duration = (24 + (14 * e.insulin / weight)) * 60;
		const peak = (duration / 2.5);
		return {
			...e,
			duration,
			peak,
		}
	});
	logger.debug('these are the last TOU: %o', lastTOU);
	const activityTOU = computeBasalActivity(lastTOU)
	logger.info('these is the total TOU activity: %o', activityTOU);

	const lastDEG = lastBasals.filter(function (e) {
		return e.drug === 'deg' || e.drug === 'Deg' || e.drug === 'tre' || e.drug === 'Tre'; // keep only the degs from the last 45 hours
	}).map(e => {
		const duration = 42 * 60;
		const peak = (duration / 3);
		return {
			...e,
			duration,
			peak,
		}
	});
	logger.debug('these are the last deg: %o', lastDEG);
	const activityDEG = computeBasalActivity(lastDEG);
	logger.info('these is the total deg activity: %o', activityDEG);

	const result = {
		lastDET,
		lastGLA,
		lastTOU,
		lastDEG,		
	};
	return activityDEG + activityDET + activityGLA + activityTOU;
}