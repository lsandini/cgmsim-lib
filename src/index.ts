import simulator from './CGMSIMsimulator';
import simulatorUVA from './UVAsimulator';
import downloads from './downloads';
import uploadEntries from './upload-entries';
import uploadActivity from './upload-activity';
import loadActivity from './load-activity';
import uploadNotes from './upload-notes';
import arrows from './arrows';


export { arrows, simulator, downloads, uploadEntries, uploadActivity, loadActivity, simulatorUVA, uploadNotes };
