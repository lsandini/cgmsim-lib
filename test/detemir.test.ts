import { computeBasalActivity } from '../src/computeBasalIOB';
describe('test detemir', () => {

	const detemir = (weight, treatments) => {
		const toujeoT = treatments
			.map(e => {
				const duration = (14 + (24 * e.insulin / weight)) * 60;
				const peak = (duration / 3);
				return {
					...e,
					duration,
					peak,
				}
			});
		return computeBasalActivity(toujeoT)
	}

	test('weight:80 ins:30 minutesAgo:300', () => {
		const weight = 80;
		const insulinActive = detemir(weight, [{
			insulin: 30,
			minutesAgo: 300
		}])
		expect(insulinActive).toMatchSnapshot();
	})

	test('insulin 5 min ago are less active then 40 min ago', () => {
		const insulin = 20;
		const weight = 80;

		const r5 = detemir(weight, [{
			insulin,
			minutesAgo: 5
		}])
		const r40 = detemir(weight, [{
			insulin,
			minutesAgo: 40
		}])
		expect(r5).toBeLessThan(r40);
	});
	test('peak has the greatest activity', () => {
		const insulin = 20;
		const weight = 80;

		const detemirPeakHours = ((14 + (24 * insulin / weight))) / 3
		const rBeforePeak = detemir(weight, [{
			insulin,
			minutesAgo: (detemirPeakHours * 60) + 10
		}]);
		const rPeak = detemir(weight, [{
			insulin,
			minutesAgo: (detemirPeakHours * 60)
		}])
		const rAfterPeak = detemir(weight, [{
			insulin,
			minutesAgo: (detemirPeakHours * 60) - 10
		}])
		expect(rBeforePeak).toBeLessThan(rPeak);
		expect(rAfterPeak).toBeLessThan(rPeak);
	});
})