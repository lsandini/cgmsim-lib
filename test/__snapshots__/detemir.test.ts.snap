// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test detemir weight:80 ins:30 minutesAgo:300 1`] = `0.032066191747993`;
